velox/exec/Operator.h

```diff
@@ -466,6 +469,10 @@ class SourceOperator : public Operator {
   void noMoreInput() override {
     VELOX_FAIL("SourceOperator does not support noMoreInput()");
   }
+
+  virtual UpdatableDataSourceSupplier updatableDataSource() {
```

Comment:
Why do we need this method in the Operator interface? Would you describe the overall design?

Reply:
To support for DELETE, not only the straightforward delete from table, but also the delete from table where predicates containing the subquery, the execution plan always follows the pattern something like Pipeline (Scan -> Filter -> Delete) -> Pipeline (Finish)

- Firstly, We have to scan the id column of the to-delete table in the Scan operator. 
- It will be passed through the pipeline until the Delete operator as the input to perform the deletion to the same table then. 
- Finally, because of this distributed execution plan, we have the finish phase to atomically commit or abort the work.

According the above execution plan, we are reading and writing to the same target table in the same driver in the pipeline way. 
So we need an abstraction UpdatableDataSource, which keeps the scanning interfaces (next ...) and also has the delete/update interfaces (delete/updataRows). 
It is created in the Scan operator and referneced by the Delete operator in the same driver.

That’s why we need this method in SourceOperator, and setup the reference when building the driver.

Reply:
@dingqiaoyi Qiaoyi, thank you for explaining and clarifying. I believe I understand what you are trying to do. To make it easier for other readers and contributors to understand the code, it would be nice to update the PR description to explain the use case you are enabling as well as the design you are implementing. Saying "Current design is mainly influenced by the PrestoDB, sharing most of the workflow" is not sufficient as folks may not be familiar with Presto's implementation.

It is nice to refactor TableWrite to extract logic that is reused in Delete operator, but let's make this refactoring in a separate PR. This will make it easier to review the changes.

The name of the CommitAwareOperator.h file suggests that it defines an operator, but it is not. Consider renaming or creating a base class fro TableWrite, Delete and other commit-aware operators to use.